telegram/ext/dispatcher.py

```diff
@@ -512,11 +577,13 @@ def add_error_handler(self, callback):
                 ``def callback(update: Update, context: CallbackContext)``
 
                 The error that happened will be present in context.error.
+            run_async (:obj:`bool`, optional): Whether this handlers callback should be run
+                asynchronously using :meth:`run_async`. Defaults to :obj:`False`.
 
         Note:
             See https://git.io/fxJuV for more info about switching to context based API.
         """
-        self.error_handlers.append(callback)
+        self.error_handlers[callback] = run_async
```

Comment:
Side effects of using a dict here are, both are changes in behaviour:
1. No guarantees regarding the order of error_handlers (unless it's python3.7 or even 3.8)
2. If the same handler is inserted twice, it will only be used once.

Reply:
I could change to a list of tuples (error_handler, run_async) …

Reply:
Yes, that's a good idea, just make sure that `remove_error_handler` will provide an expectable semantics (especially as now the caller would need to know what is the `run_async` value).

A twist on your proposal would be (not sure if it really improves):
Wrap the callback into our own object and make it `__eq__` and `__hash__` only on the callback. Something like:
```python
class _PtbErrorHandler:

    def __init__(self, error_handler: Callable, run_async: bool):
       self._cb = error_handler
       self._is_async = run_async

    def __hash__(self):
       return hash(self._cb)

    def __eq__(self, other):
       return self._cb == other._cb

    # And we can even take it a step further (totally optional):
    def __call__(self, *args, **kwargs):
        if self._is_async:
            async_execution(self._cb, *args, **kwargs)
        else:
            self._cb(*args, **kwargs)
```

That might help other parts of the code seamlessly handle the changes of how we internally keep the callback.

Reply:
Mh … If we go that route, we might as well set up a full blown handler logic for error handlers … Currently, `remove(error_handler)` will just remove the first occurrence of `error_handler` from the list. I'd just do the same for the duples, ignoring the `run_async` part (ofc documenting that). IMHO, if someone registers the same error_handler twice or worse, once synchronously and once asynchronously, they don't deserve us special casing for that …

Reply:
Ok. I agree, you may resolve this conversation.